apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,235 @@
+      return req.user;
+    }
+    if (req.headers?.authorization?.length) {
+      const token = req.headers.authorization.split(' ')[1];
+      if (token) {
+        return jwt.decode(token);
+      }
+    }
+
+    return null;
+  }
+
+  private getCacheKey(context: ExecutionContext): string {
+    const { organizationId } = this.getReqUser(context) || {};
+    const env = process.env.NODE_ENV;
+
+    return `${env}-${organizationId}-${this.getIdempotencyKey(context)}`;
+  }
+
+  async setCache(
+    key: string,
+    val: { status: ReqStatusEnum; bodyHash: string; data?: any; statusCode?: number },
+    ttl: number,
+    ifNotExists?: boolean
+  ): Promise<string | null> {
+    try {
+      if (ifNotExists) {
+        return await this.cacheService.setIfNotExist(key, JSON.stringify(val), { ttl });
+      }
+      await this.cacheService.set(key, JSON.stringify(val), { ttl });
+    } catch (err) {
+      Logger.warn(`An error occurred while setting idempotency cache, key:${key} error: ${err.message}`, LOG_CONTEXT);
+    }
+
+    return null;
+  }
+
+  private buildError(error: any): HttpException {
+    const statusCode = error.status || error.response?.statusCode || 500;
+    if (statusCode == 500 && !error.response) {
+      //some unhandled exception occured
+      return new InternalServerErrorException();
+    }
+
+    return new HttpException(error.response || error.message, statusCode, error.response?.options);
+  }
+
+  private setHeaders(response: any, headers: Record<string, string>) {
+    Object.keys(headers).map((key) => {
+      if (headers[key]) {
+        response.set(key, headers[key]);
+      }
+    });
+  }
+
+  private hashRequestBody(body: object): string {
+    const hash = createHash('blake2s256');
+    hash.update(Buffer.from(JSON.stringify(body)));
+
+    return hash.digest('hex');
+  }
+
+  private async handlerDuplicateRequest(context: ExecutionContext, bodyHash: string): Promise<Observable<any>> {
+    const cacheKey = this.getCacheKey(context);
+    const idempotencyKey = this.getIdempotencyKey(context)!;
+    const data = await this.cacheService.get(cacheKey);
+    this.setHeaders(context.switchToHttp().getResponse(), { [HEADER_KEYS.IDEMPOTENCY]: idempotencyKey });
+    const parsed = JSON.parse(data);
+    if (parsed.status === ReqStatusEnum.PROGRESS) {
+      // api call is in progress, so client need to handle this case
+      Logger.error(`previous api call in progress rejecting the request. key:${idempotencyKey}`, LOG_CONTEXT);
+      const error = 'IDEMPOTENCY_REQUEST_PROCESSING';
+      this.setHeaders(context.switchToHttp().getResponse(), {
+        [HEADER_KEYS.IDEMPOTENCY_CONFLICT]: error,
+        [HEADER_KEYS.RETRY_AFTER]: `1`,
+      });
+
+      return throwError(
+        () =>
+          new HttpException(
+            {
+              error,
+              message: `request ${idempotencyKey} is currently being processed. Please retry after 1 second`,
+            },
+            429
+          )
```

Comment:
Small nit for better DevX, and removing `error` and moving to `409` per https://github.com/novuhq/novu/pull/4531#issuecomment-1766557580.

```suggestion
          new ConflictException(`Request with key "${idempotencyKey}" is currently being processed. Please retry after 1 second`)
```

Reply:
@Cliftonz  if change to 409

1. we cant [really send retry-after](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) header
2. more importantly client libraries(not novu client but the http client) also retry only when 429 with  Retry-After header is sent, so we might have to put more effort into client sdks

Reply:
I agree with you on this, our initial idea was to use 409, but because of the effort that would involve in client sdks we thought of tweaking the the server, but if we are okay to do the heavy lifting of modifying the http clients then going with 409 is better as it aligns with the specs. 
now the `Retry-After` is not mentioned in the specs, since its exponential retry we could possibly remove this as well.

Reply:
Ok great, yes we are ok to do some heavy lifting on the client sdks to be true to the specification.

I agree, let's also keep the `retry-after` header as part of this implementation.